### PR TITLE
Add support of importing modules from packages

### PR DESCRIPTION
**Pull request by Arthapz**
This is a Proof of concepts, may need a lot of optimisation / refinements

the module distribution files are based on https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p2473r1.pdf

### CODE REVIEW

File: xmake/rules/c++/modules/modules_support/common.lua [hunk 1]
```diff
@@ -100,6 +100,64 @@ function patch_sourcebatch(target, sourcebatch)
...
 
+-- extract packages modules dependencies
+function get_all_package_modules(target, modules, opt)
+    local package_modules
+
+    -- parse all meta-info and append their informations to the package store
+    for name, package in pairs(target:pkgs()) do
+        package_modules = package_modules or {}
+        local modules_dir = path.join(package:installdir(), "modules", name)
+        local meta_files = os.match(path.join(modules_dir, "**.meta-info"), false)
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/common.lua:111]: os.files

**Comment by waruqi** [comment]
Does it work now? Can you provide an example project with package? and how to test it?

**Comment by Arthapz** [comment]
this PR works on the small tests i did, but it need furthermore test to ensures it works with more complex project
and for unitary tests i don't know how to test it, maybe add a fake test package to xrepo (or a private repository) that contains modules ?

**Comment by waruqi** [comment]
> this PR works on the small tests i did, but it need furthermore test to ensures it works with more complex project and for unitary tests i don't know how to test it, maybe add a fake test package to xrepo (or a private repository) that contains modules ?

Add a test project and define a local package with modules in this project directly. 

then call `package.tools.xmake` to install it and use `add_requires/add_packages` to use it.

see https://xmake.io/#/package/local_3rd_source_library

**Comment by Arthapz** [comment]
are the package installation handled differently between local and external repo ? 

```lua
-- install modules
rule("c++.build.modules.install")
    set_extensions(".mpp", ".mxx", ".cppm", ".ixx")

    before_install(function (target)
        import("modules_support.common")

        -- we cannot use target:data("cxx.has_modules"),
        -- because on_config will be not called when installing targets
        print("TEST1")
        if common.contains_modules(target) then
            print("TEST2")
            common.install_module_target(target)
        end
    end)
```

the install rules doesn't seem to be called when installing local package

```
[100%]: build ok!
xmake install -y -o C:\Users\arthu\development\repos\xmake\tests\projects\c++\modules\packages\build\.packages\b\bar\latest\8574311f4b544fa89a3d8c3ee79d1929 -P C:\Users\arthu\development\repos\xmake\tests\projects\c++\modules\packages\my-repo\packages\b\bar\src --diagnosis --verbose --yes
installing bar ..
installing bar to C:\Users\arthu\development\repos\xmake\tests\projects\c++\modules\packages\build\.packages\b\bar\latest\8574311f4b544fa89a3d8c3ee79d1929 ..
> copy ..\..\..\..\..\build\.packages\b\bar\latest\cache\build_8574311f\windows\x64\release\bar.lib to C:\Users\arthu\development\repos\xmake\tests\projects\c++\modules\packages\build\.packages\b\bar\latest\8574311f4b544fa89a3d8c3ee79d1929\lib
> copy bar.mpp to C:\Users\arthu\development\repos\xmake\tests\projects\c++\modules\packages\build\.packages\b\bar\latest\8574311f4b544fa89a3d8c3ee79d1929\modules\bar\bar.mpp
install ok!
finding bar from xmake ..
checking for xmake::bar ... bar latest
```

**Comment by waruqi** [comment]
it works for me 

```console
xmake install -y -o /mnt/xmake/tests/projects/c++/modules/packages/build/.packages/b/bar/latest/4e0143c97b65425b855ad5fd03038b6a -P /mnt/xmake/tests/projects/c++/modules/packages/my-repo/packages/b/bar/src --diagnosis --verbose
TEST1.  <--------
TEST2
installing bar ..
installing bar to /mnt/xmake/tests/projects/c++/modules/packages/build/.packages/b/bar/latest/4e0143c97b65425b855ad5fd03038b6a ..
> copy ../../../../../build/.packages/b/bar/latest/cache/build_4e0143c9/linux/x86_64/release/libbar.a to /mnt/xmake/tests/projects/c++/modules/packages/build/.packages/b/bar/latest/4e0143c97b65425b855ad5fd03038b6a/lib
> copy bar.mpp to /mnt/xmake/tests/projects/c++/modules/packages/build/.packages/b/bar/latest/4e0143c97b65425b855ad5fd03038b6a/modules/bar/bar.mpp
install ok!
```

But it does not build project for gcc.

```
/usr/bin/gcc -c -m64 -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -isystem /mnt/xmake/tests/projects/c++/modules/packages/build/.packages/f/foo/latest/4e0143c97b65425b855ad5fd03038b6a/include -isystem /mnt/xmake/tests/projects/c++/modules/packages/build/.packages/b/bar/latest/4e0143c97b65425b855ad5fd03038b6a/include -fmodules-ts -fmodule-mapper=build/packages/mapper.txt -D_GLIBCXX_USE_CXX11_ABI=0 -DNDEBUG -o build/.objs/packages/linux/x86_64/release/src/main.cpp.o src/main.cpp
checking for flags (-MMD -MF) ... ok
checking for flags (-fdiagnostics-color=always) ... ok
error: src/main.cpp:1:1: error: unknown Compiled Module Interface: no such module
    1 | import foo;
      | ^~~~~~
src/main.cpp:2:1: error: unknown Compiled Module Interface: no such module
    2 | import bar;
      | ^~~~~~
In module imported at src/main.cpp:1:1:
foo: error: failed to read compiled module: Unknown CMI mapping
foo: note: imports must be built before being imported
foo: fatal error: returning to the gate for a mechanical issue
compilation terminated.
```

**Comment by Arthapz** [comment]
> it works for me
> 
> ```
> xmake install -y -o /mnt/xmake/tests/projects/c++/modules/packages/build/.packages/b/bar/latest/4e0143c97b65425b855ad5fd03038b6a -P /mnt/xmake/tests/projects/c++/modules/packages/my-repo/packages/b/bar/src --diagnosis --verbose
> TEST1.  <--------
> TEST2
> installing bar ..
> installing bar to /mnt/xmake/tests/projects/c++/modules/packages/build/.packages/b/bar/latest/4e0143c97b65425b855ad5fd03038b6a ..
> > copy ../../../../../build/.packages/b/bar/latest/cache/build_4e0143c9/linux/x86_64/release/libbar.a to /mnt/xmake/tests/projects/c++/modules/packages/build/.packages/b/bar/latest/4e0143c97b65425b855ad5fd03038b6a/lib
> > copy bar.mpp to /mnt/xmake/tests/projects/c++/modules/packages/build/.packages/b/bar/latest/4e0143c97b65425b855ad5fd03038b6a/modules/bar/bar.mpp
> install ok!
> ```
> 
> But it does not build project for gcc.
> 
> ```
> /usr/bin/gcc -c -m64 -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -isystem /mnt/xmake/tests/projects/c++/modules/packages/build/.packages/f/foo/latest/4e0143c97b65425b855ad5fd03038b6a/include -isystem /mnt/xmake/tests/projects/c++/modules/packages/build/.packages/b/bar/latest/4e0143c97b65425b855ad5fd03038b6a/include -fmodules-ts -fmodule-mapper=build/packages/mapper.txt -D_GLIBCXX_USE_CXX11_ABI=0 -DNDEBUG -o build/.objs/packages/linux/x86_64/release/src/main.cpp.o src/main.cpp
> checking for flags (-MMD -MF) ... ok
> checking for flags (-fdiagnostics-color=always) ... ok
> error: src/main.cpp:1:1: error: unknown Compiled Module Interface: no such module
>     1 | import foo;
>       | ^~~~~~
> src/main.cpp:2:1: error: unknown Compiled Module Interface: no such module
>     2 | import bar;
>       | ^~~~~~
> In module imported at src/main.cpp:1:1:
> foo: error: failed to read compiled module: Unknown CMI mapping
> foo: note: imports must be built before being imported
> foo: fatal error: returning to the gate for a mechanical issue
> compilation terminated.
> ```

yes i spoke too fast

**Comment by Arthapz** [comment]
but any idea why the package test build when running `xmake -vD --yes` but not when running `xmake --yes` ?

**Comment by Arthapz** [comment]
Ok it's this function that fail on msvc without -vD (when build a package)

```
function get_ifcoutputflag(target)
    local ifcoutputflag = _g.ifcoutputflag
    if ifcoutputflag == nil then
        local compinst = target:compiler("cxx")
        if compinst:has_flags("-ifcOutput", "cxxflags", {flagskey = "cl_ifc_output"})  then
            ifcoutputflag = "-ifcOutput"
        end
        assert(ifcoutputflag, "compiler(msvc): does not support c++ module!")
        _g.ifcoutputflag = ifcoutputflag or false
    end
    return ifcoutputflag or nil
end
```

**Comment by waruqi** [comment]
you can pass `-vD` in [test_headerunits.lua](https://github.com/xmake-io/xmake/pull/3228/files#diff-d5126bd1bcf1689011902e884616e08ac8ed11fb396ce798e3f768c601272d84)

to see verbose checkflags logs

**Comment by Arthapz** [comment]
when adding -vD the test pass :/ i don't know how to debug this

**Comment by waruqi** [comment]
> when adding -vD the test pass :/ i don't know how to debug this

I don't know, I need some time to debug it.

maybe need debug 
https://github.com/xmake-io/xmake/blob/master/xmake/modules/detect/tools/cl/has_flags.lua
https://github.com/xmake-io/xmake/blob/master/xmake/modules/lib/detect/has_flags.lua

**Comment by waruqi** [comment]
> Ok it's this function that fail on msvc without -vD (when XMake is trying to build a package)
> 
> ```lua
> function get_ifcoutputflag(target)
>     local ifcoutputflag = _g.ifcoutputflag
>     if ifcoutputflag == nil then
>         local compinst = target:compiler("cxx")
>         if compinst:has_flags("-ifcOutput", "cxxflags", {flagskey = "cl_ifc_output"})  then
>             ifcoutputflag = "-ifcOutput"
>         end
>         assert(ifcoutputflag, "compiler(msvc): does not support c++ module!")
>         _g.ifcoutputflag = ifcoutputflag or false
>     end
>     return ifcoutputflag or nil
> end
> ```

Because of some incorrect flags detection (missing input arguments), when calling `cl /ifcOutput -Foxxx.obj test.cpp` to test for flags, `-Foxxx.obj` is incorrectly used as an input argument to `/ifcOutput`. cl generates xxx.obj to the default path. When there are multiple parallel tasks, access to this obj file is taken up and the test fails.

Some other msvc flags detections have the same problem.

**Comment by waruqi** [comment]
It work now, I have fixed it.

**Comment by waruqi** [comment]
I tested for gcc/clang on my machine, it works now.

File: xmake/rules/c++/modules/modules_support/common.lua [hunk 2]
```diff
@@ -102,6 +102,92 @@ function patch_sourcebatch(target, sourcebatch)
...
+                table.append(needed_modules, required)
+            end
+        end
+    end
+
+    -- append all package dependencies
+    local module_names = {}
+    for name, _ in pairs(package_modules_data) do
+        table.append(module_names, name)
+    end
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/common.lua:172]: why not use `local module_names = table.keys(package_modules_data)`?

File: xmake/rules/c++/modules/modules_support/common.lua [hunk 3]
```diff
@@ -102,6 +102,92 @@ function patch_sourcebatch(target, sourcebatch)
...
+
+    local culled
+    for i, name in ipairs(module_names) do
+        culled = culled or {}
+        if table.find(needed_modules, name) and package_modules_data[name] and not culled[name] then
+            culled[name] = package_modules_data[name]
+
+            -- table.remove(module_names, i)
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/common.lua:180]: remove unused code

File: xmake/rules/c++/modules/modules_support/common.lua [hunk 4]
```diff
@@ -102,6 +102,92 @@ function patch_sourcebatch(target, sourcebatch)
...
+    end
+
+    -- append all package dependencies
+    local module_names = {}
+    for name, _ in pairs(package_modules_data) do
+        table.append(module_names, name)
+    end
+
+    local culled
+    for i, name in ipairs(module_names) do
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/common.lua:175]: i -> _

File: xmake/rules/c++/modules/xmake.lua [hunk 5]
```diff
@@ -67,6 +67,24 @@ rule("c++.build.modules.builder")
...
 
+            -- extract packages modules dependencies
+            local package_modules_data = common.get_all_package_modules(target, modules, opt)
+            if package_modules_data then
+                -- cull unused modules
+                package_modules_data = common.cull_unused_modules(target, modules, package_modules_data)
+                if package_modules_data then
+                    -- append to sourcebatch
+                    for name, package_module_data in pairs(package_modules_data) do
+                        table.append(sourcebatch.sourcefiles, package_module_data.file)
```
Comment (waruqi) [inline xmake/rules/c++/modules/xmake.lua:78]: you should use `table.insert`

File: xmake/rules/c++/modules/modules_support/common.lua [hunk 6]
```diff
@@ -102,6 +102,92 @@ function patch_sourcebatch(target, sourcebatch)
...
+end
+
+-- cull unused modules
+function cull_unused_modules(target, modules, package_modules_data)
+    local needed_modules = {}
+    -- append all target dependencies
+    for _, module in pairs(modules) do
+        if module.requires then
+            for required, _ in pairs(module.requires) do
+                table.append(needed_modules, required)
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/common.lua:163]: please use table.insert, `table.append` only for appending table

**Comment by waruqi** [comment]
> This is a Proof of concepts, may need a lot of optimisation / refinements
> 
> the module distribution files are based on https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p2473r1.pdf

I sent this pr experimental support for c++ modules distribution to sg15@lists.isocpp.org and got some feedback.

> This is a great first step. Thanks for pushing this.
> 
> I should note, however, that the proposal on that paper didn't reach consensus.
> 
> The approach that did reach consensus, later, was the one proposed in https://wg21.link/P2577R2 with the specific convention for the translation in https://wg21.link/P2701R0
> 
> Those have already been incorporated into the draft modules experience TR, which you can see the source in https://github.com/cplusplus/modules-ecosystem-tr/blob/master/appendix/discovery-in-prebuilt-library-releases.tex and a recent enough rendering on https://github.com/cplusplus/modules-ecosystem-tr/files/10071184/iso_cpp_modules_ecosystem_technical_report.pdf
> 
> Note that we haven't yet reached a consensus on the suffix for the file name containing the metadata, nor on the format of the metadata file itself.
> 
> Both still require additional papers and experiments.

**Pull request merged**